java/cuvs-java/src/main/java22/com/nvidia/cuvs/internal/BruteForceIndexImpl.java

```diff
@@ -158,25 +159,28 @@ private IndexReference build(
 
     try (var resourcesAccessor = resources.access()) {
       long cuvsResources = resourcesAccessor.handle();
-      MemorySegment datasetMemorySegmentP = allocateRMMSegment(cuvsResources, datasetBytes);
+      try (var closeableDataMemorySegmentP = allocateRMMSegment(cuvsResources, datasetBytes)) {
+        MemorySegment datasetMemorySegmentP = closeableDataMemorySegmentP.handle();
 
-      cudaMemcpy(datasetMemorySegmentP, datasetMemSegment, datasetBytes, INFER_DIRECTION);
+        cudaMemcpy(datasetMemorySegmentP, datasetMemSegment, datasetBytes, INFER_DIRECTION);
 
-      long[] datasetShape = {rows, cols};
-      var tensorDataArena = Arena.ofShared();
-      MemorySegment datasetTensor =
-          prepareTensor(tensorDataArena, datasetMemorySegmentP, datasetShape, 2, 32, 2, 1);
+        long[] datasetShape = {rows, cols};
+        var tensorDataArena = Arena.ofShared();
+        MemorySegment datasetTensor =
+            prepareTensor(tensorDataArena, datasetMemorySegmentP, datasetShape, 2, 32, 2, 1);
 
-      var returnValue = cuvsStreamSync(cuvsResources);
-      checkCuVSError(returnValue, "cuvsStreamSync");
+        var returnValue = cuvsStreamSync(cuvsResources);
+        checkCuVSError(returnValue, "cuvsStreamSync");
 
-      returnValue = cuvsBruteForceBuild(cuvsResources, datasetTensor, 0, 0.0f, index);
-      checkCuVSError(returnValue, "cuvsBruteForceBuild");
+        returnValue = cuvsBruteForceBuild(cuvsResources, datasetTensor, 0, 0.0f, index);
+        checkCuVSError(returnValue, "cuvsBruteForceBuild");
 
-      returnValue = cuvsStreamSync(cuvsResources);
-      checkCuVSError(returnValue, "cuvsStreamSync");
+        returnValue = cuvsStreamSync(cuvsResources);
+        checkCuVSError(returnValue, "cuvsStreamSync");
 
-      return new IndexReference(datasetMemorySegmentP, datasetBytes, tensorDataArena, index);
+        return new IndexReference(
+            closeableDataMemorySegmentP.release(), datasetBytes, tensorDataArena, index);
```

Comment:
I still lean more towards an explicit try-catch --  IDK, this pattern looks a bit strange in Java. I understand it because I have a C/C++ background, but I bet it's not immediate for "pure Java" folks. Then again, people that are going to work on this library would need to have some C/C++ knowledge, so...

like this much better than the original though; I would even go an extra mile here (optional and/or good for a follow-up PR) and make `IndexReference` taking a `CloseableRMMAllocation` as the first param (, changing/moving the logic of cleaning up in `destroyIndex` so it calls `close` on the `CloseableRMMAllocation`.

Reply:
If you do this, you can even avoid exposing release:
```
IndexReference(
    new CloseableRMMAllocation(closeableDataMemorySegmentP), // note that we have a "copy-ctor" here, not calling release
    datasetBytes, 
    tensorDataArena, 
    index);

```
and in `CloseableRMMAllocation`
```
public CloseableRMMAllocation(CloseableRMMAllocation other) {
   var rawPointer = other.release(); // release can be private! or inlined here
   // copy other private members as needed too
}
```
So you cannot just release a raw pointer, you need to transfer it to another `CloseableRMMAllocation`, making it explicit you need to `close` it (and Intellij should help you with warnings too if you don't do that).

Wdyt?